### PR TITLE
Model training sequence

### PR DESCRIPTION
**Pull request by RayNawfal**
Build Training model Sequence with functions to:
- download checkpoint from S3 to /temp dir
- unzip checkpoint from /temp dir to /temp_unzip
- zip checkpoint from /temp_unzip to /temp dir
- upload new checkpoint to S3

Clickup link:
https://app.clickup.com/t/24308805/VDM-1450

### CODE REVIEW

File: audio-training-seq/model-training-sequence/main.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+
+import os
+import boto3
+#import tarfile
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py]: leftover?

File: audio-training-seq/model-training-sequence/main.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+def download_object(aws_key, aws_secret, bucket, object_name, downloaded_filename):
+    session = boto3.session.Session()
+    s3_client = session.client(
+        service_name='s3',
+        aws_access_key_id=aws_key,
+        aws_secret_access_key=aws_secret,
+        )   
+    with open(downloaded_filename, "wb") as file:
+        response = s3_client.download_fileobj(bucket, object_name, file)
+        print(f"Object file downloaded. Error code: {response}")
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py:42]: Why `Error code`? Perhaps it should be `Status/Response code`? 
Also line 57.

File: audio-training-seq/model-training-sequence/main.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    try:
+        dir = os.makedirs(desired_dir)
+    except FileExistsError:
+        pass
+    with tarfile.open(source_dir, "r") as tar:
+        tar.extractall(desired_dir)
+    print(f"Files extracted from tarfile completed...")
+
+
+async def run(context, input, arg1, arg2, arg3, arg4):
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py]: instead of arg1, arg2 etc. I would just do as we always do:
```py
async def run(context, input, key, secret, bucket, object):
```

File: audio-training-seq/model-training-sequence/main.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    key = arg1
+    secret = arg2
+    bucket = arg3
+    print(f"STARTING THE SEQUENCE")
+    lst = checkpoint_search(key, secret, bucket)
+    
+    #create directory if not existing
+    try:
+        path = "temp"
+        dir = os.makedirs(path)
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py]: I believe this will never throw an error. `Run()` function is the first one to be runu by Runner, and there is no code before it that would create temp directory before it. Unless you upload the Sequence with such.
Alternatively you could use:
```py
 os.makedirs(path, exist_ok=True)
```
which is shorter :D

File: audio-training-seq/model-training-sequence/main.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    except FileExistsError:
+        pass
+
+    object = arg4
+    download_path = os.path.join(path, object) 
+    download_object(key, secret, bucket, object, download_path)
+
+    # Unzip tarfile
+    file_path = "/".join(["temp", arg4])
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py]: Instead of `temp` you could use `path` variable.

File: audio-training-seq/model-training-sequence/main.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
+
+import os
+import boto3
+#import tarfile
+import asyncio
+from scramjet import streams
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py:6]: Use a longer import, but a shorter return!
```py
from scramjet.streams import Stream
```
and then line 122.

File: audio-training-seq/model-training-sequence/main.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    # Upload the new checkpoint tarfile
+    up_path = "temp/checkpoint.tar.gz"
+    upload_name = "checkpoint.tar.gz"
+    upload_object(key, secret, bucket, up_path, upload_name)
+
+    # list the new objects found on S3
+    bucket_objects = checkpoint_search(key, secret, bucket)
+
+    return streams.Stream.read_from(f"{bucket_objects}\n")
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py:117]:
```py
return Stream.read_from(f"{bucket_objects}\n")
```

File: audio-training-seq/functions.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+def checkpoint_search(aws_key, aws_secret, bucket):
+    session = boto3.session.Session()
+    s3_client = session.client(
+        service_name='s3',
+        aws_access_key_id=aws_key,
+        aws_secret_access_key=aws_secret,
+
+        )
+    response = s3_client.list_objects(Bucket=bucket)
+    print(response['ResponseMetadata']['HTTPStatusCode'])
```
Comment (S4adam) [inline audio-training-seq/functions.py:21]: Could this raise `KeyError`?

File: audio-training-seq/model-training-sequence/main.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        aws_secret_access_key=aws_secret,
+
+        )
+    response = s3_client.list_objects(Bucket=bucket)
+    print(response['ResponseMetadata']['HTTPStatusCode'])
+
+    bucket_lst = []
+    if 'Contents' in response:
+        print("Objects found in the bucket:")
+        for obj in response['Contents']:
```
Comment (S4adam) [inline audio-training-seq/model-training-sequence/main.py:23]: To be safe with possible `KeyError` exception, you could use try/except instead:
```py
    try:
        for obj in response['Contents']:
            bucket_lst.append(obj['Key'])
        print("Objects found in the bucket:")
    except KeyError:
        return f"No objects found in the bucket"
```

File: audio-training-seq/train-predict.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+plt.xlabel('Epoch')
+plt.ylabel('Accuracy [%]')
+
+model.save("model-01.keras")
+print("model has been saved")
+
+path_model = "PATH/TO/model-01.keras"
+path_audio = "PATH/TO/WAV/FILE"
+
+model = tf.keras.models.load_model(path_model)
```
Comment (S4adam) [inline audio-training-seq/train-predict.py:136]: try/catch?

File: audio-training-seq/train-predict.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    layers.Dropout(0.5),
+    layers.Dense(64, activation='relu'),
+    layers.Dense(32, activation='relu'),
+    layers.Dense(10, activation='softmax'),
+])
+
+model.summary()
+
+model.compile(
+    optimizer=tf.keras.optimizers.Adam(), #learning_rate=0.001),
```
Comment (S4adam) [inline audio-training-seq/train-predict.py:107]: I know that guy ^-^

File: audio-training-seq/train-predict.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+           pass
+        path_folder = os.path.join(path, folder)
+        for file in os.listdir(path_folder):
+            dir_index = path_dataset.index(folder)
+            audio_dir = os.path.join(path_folder, file)
+            spectrogram = get_spectrogram(audio_dir) 
+            yield spectrogram, dir_index
+
+
+dataset_path = "PATH/TO/FOLDER/mini_speech_commands"
```
Comment (S4adam) [inline audio-training-seq/train-predict.py:62]: hmmm, could you move it up, and adjust the path?

File: audio-training-seq/train-predict.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+def get_record(path):
+    path_dataset = os.listdir(path)
+    for folder in path_dataset:
+        label = folder
+        print(label.decode("utf-8"))
+        label = label.decode("utf-8")
+        if label not in label_names:
+           label_names.append(label)
+        else:
+           pass
```
Comment (S4adam) [inline audio-training-seq/train-predict.py:53]: if `else pass` is to be left here, then remove it

**Review by S4adam — COMMENTED** [review]
This is so goood and will be impactful just like Raspberry was for Scramjet. 
There are no errors I believe. Aside from the comments, I would improve some variable names and move **_some_** hard coded values to the top - but then again, it will be a lot of values which most likely would qualify for using a separate config file so that the code doesn't become even longer.

File: audio-training-seq/model-training-sequence/main.py (hunk 2 above)
Comment (RayNawfal) [reply audio-training-seq/model-training-sequence/main.py:42]: Response don't include HTTP status code 200 or 202 but rather will throw an error code if it occurs

File: audio-training-seq/train-predict.py (hunk 11 above)
Comment (RayNawfal) [reply audio-training-seq/train-predict.py:107]: very useful when needed --> stochastic gradient descent method

File: audio-training-seq/train-predict.py (hunk 12 above)
Comment (RayNawfal) [reply audio-training-seq/train-predict.py:62]: test sample to be used in other files

File: audio-training-seq/train-predict.py (hunk 13 above)
Comment (RayNawfal) [reply audio-training-seq/train-predict.py:53]: temporary not to raise any errors.

**Review by S4adam — APPROVED** [review]
